### PR TITLE
Show phrase structure in editor

### PR DESCRIPTION
**Pull request by markATavail**
Show phrase structure in editor.  This will need to be rebased after @richATavail 's PR merges, but I think the conflicts will be very minor.
[Edit 2023-01-10: I've rebased onto the latest development and force-pushed mvg_development.  Rich's code review comments are probably inaccessible now, but all issues he pointed out have allready been resolved.]

### CODE REVIEW

File: avail/src/main/kotlin/avail/compiler/splitter/MessageSplitter.kt [hunk 1]
```diff
@@ -1064,16 +1249,27 @@ constructor(messageName: A_String) {
...
+			group = group.apply {
+				Group(
+					startInName,
+					currentPositionForEnd,  // Include the '~'.
+					beforeDagger.applyCaseInsensitive(),
+					hasDagger,
+					afterDagger.applyCaseInsensitive(),
+					maximumCardinality)
+			}
 		}
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/MessageSplitter.kt:1262]: Are you trying to overwrite `group` with the contents in the `apply` block? 
The `Group` constructed  inside the `apply` block is just being thrown away and `group` preserves its previous value.

File: avail/src/main/kotlin/avail/compiler/splitter/MessageSplitter.kt [hunk 2]
```diff
@@ -1262,7 +1484,28 @@ constructor(messageName: A_String) {
 
 	override fun toString() = buildString { dumpForDebug(this) }
 
-	companion object {
+	companion object
+	{
+		/**
+		 * The public factory for splitting message names.  The result may be
+		 * cached in an [LRUCache].  Since the [messageName] is used as a key in
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/MessageSplitter.kt:1491]: You reference the value parameter, `messageName`, in a document link but you do not provide an `@param` comment description for `messageName`. When clicked through in the IDE, it provides no information other than type information.

File: avail/src/main/kotlin/avail/compiler/splitter/MessageSplitterTokenizer.kt [hunk 3]
```diff
@@ -87,7 +88,7 @@ class MessageSplitterTokenizer
 	private val messageNameSize: Int = messageName.tupleSize
 
 	/** The current one-based index into the messageName. */
-	private var positionInName = 1
+	private var startInName = 1
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/MessageSplitterTokenizer.kt:91]: The field name has changed which implies its usage/purpose has changed, however the comment has not been updated. Is the comment still accurate?

File: avail/src/main/kotlin/avail/compiler/splitter/NumberedChoice.kt [hunk 4]
```diff
@@ -80,8 +80,11 @@ import java.util.Collections
  * @param alternation
  *   The enclosed [Alternation].
  */
-internal class NumberedChoice constructor(private val alternation: Alternation)
-: Expression(alternation.positionInName)
+internal class NumberedChoice constructor(
+	startInName: Int,
+	pastEndInName: Int,
+	private val alternation: Alternation,
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/NumberedChoice.kt:86]: Missing constructor comments.

File: avail/src/main/kotlin/avail/compiler/splitter/Optional.kt [hunk 5]
```diff
@@ -77,14 +77,16 @@ import java.util.Collections
  *
  * Construct a new `Optional`.
  *
- * @param positionInName
+ * @param startInName
  *   The position of the group or token in the message name.
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/Optional.kt:81]: Another place where the parameter name has changed but the comment describing it has not. Is this just a more accurate name or does the comment need to be updated?

File: avail/src/main/kotlin/avail/compiler/splitter/RawKeywordTokenArgument.kt [hunk 6]
```diff
@@ -54,16 +54,17 @@ import avail.descriptor.types.A_Type
...
  * @param absoluteUnderscoreIndex
  *   The one-based index of this argument within the entire message name's list
  *   of arguments.
  */
 internal class RawKeywordTokenArgument constructor(
-	positionInName: Int,
-	absoluteUnderscoreIndex: Int)
-: RawTokenArgument(positionInName, absoluteUnderscoreIndex)
+	startInName: Int,
+	pastEndInName: Int,
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/RawKeywordTokenArgument.kt:65]: missing constructor comment

File: avail/src/main/kotlin/avail/compiler/splitter/RawLiteralTokenArgument.kt [hunk 7]
```diff
@@ -55,16 +55,17 @@ import avail.descriptor.types.A_Type
...
  * @param absoluteUnderscoreIndex
  *   The one-based index of this argument within the entire message name's list
  *   of arguments.
  */
 internal class RawLiteralTokenArgument constructor(
-	positionInName: Int,
-	absoluteUnderscoreIndex: Int)
-: RawTokenArgument(positionInName, absoluteUnderscoreIndex)
+	startInName: Int,
+	pastEndInName: Int,
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/RawLiteralTokenArgument.kt:66]: missing constructor comment

File: avail/src/main/kotlin/avail/compiler/splitter/RawTokenArgument.kt [hunk 8]
```diff
@@ -51,16 +51,17 @@ import avail.descriptor.types.A_Type
...
  * @param absoluteUnderscoreIndex
  *   The one-based index of this argument within the entire message name's list
  *   of arguments.
  */
 internal open class RawTokenArgument constructor(
-	positionInName: Int,
-	absoluteUnderscoreIndex: Int)
-: Argument(positionInName, absoluteUnderscoreIndex)
+	startInName: Int,
+	pastEndInName: Int,
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/RawTokenArgument.kt:62]: missing constructor comment

File: avail/src/main/kotlin/avail/compiler/splitter/SectionCheckpoint.kt [hunk 9]
```diff
@@ -62,14 +62,16 @@ import avail.descriptor.types.A_Type
...
+ * @param startInName
  *   The section checkpoint's position in the message name.
  * @param subscript
  *   This section checkpoint's one-based index.
  */
 internal class SectionCheckpoint constructor(
-	positionInName: Int,
-	private val subscript: Int) : Expression(positionInName)
+	startInName: Int,
+	pastEndInName: Int,
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/SectionCheckpoint.kt:72]: missing constructor comment

File: avail/src/main/kotlin/avail/compiler/splitter/Sequence.kt [hunk 10]
```diff
@@ -84,25 +84,26 @@ import avail.exceptions.SignatureException
...
+ * @param startInName
  *   The position that this `Sequence` begins within the method name.
+ * @param expressions
+ *   The [List] of [Expression]s that I comprise.
  */
 internal class Sequence constructor(
-	positionInName: Int
-) : Expression(positionInName)
+	startInName: Int,
+	pastEndInName: Int,
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/Sequence.kt:94]: missing constructor comment

File: avail/src/main/kotlin/avail/compiler/splitter/Simple.kt [hunk 11]
```diff
@@ -55,14 +56,24 @@ import avail.descriptor.types.A_Type
...
- * @param positionInName
+ * @param startInName
  *   The one-based index of the token within the entire name string.
+ * @param pastEndInName
+ *   After the [MessageSplitter] has produced a complete [Expression] tree, each
+ *   node will have its [pastEndInName] populated.
  */
 internal class Simple constructor(
+	startInName: Int,
+	pastEndInName: Int,
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/Simple.kt:67]: missing constructor comment

File: avail/src/main/kotlin/avail/compiler/splitter/VariableQuote.kt [hunk 12]
```diff
@@ -57,16 +57,17 @@ import java.util.IdentityHashMap
...
  * @param absoluteUnderscoreIndex
  *   The one-based index of this argument within the entire message name's list
  *   of arguments.
  */
 internal class VariableQuote constructor(
-	positionInName: Int,
-	absoluteUnderscoreIndex: Int)
-: Argument(positionInName, absoluteUnderscoreIndex)
+	startInName: Int,
+	pastEndInName: Int,
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/compiler/splitter/VariableQuote.kt:68]: missing constructor comment

File: avail/src/main/kotlin/avail/descriptor/phrases/SendPhraseDescriptor.kt [hunk 13]
```diff
@@ -241,9 +255,17 @@ class SendPhraseDescriptor private constructor(
...
 		 * which match parts of the actual text of the message name).
 		 *
-		 * @param tokens
+		 * @param consumedTokens
 		 *   The [tuple][A_Tuple] of [tokens][A_Token] that comprise the
 		 *   [send][SendPhraseDescriptor].
+		 * @param consumedTokens
+		 *   The list of all tokens collected for this send phrase.  This includes
+		 *   only those tokens that are operator or keyword tokens that correspond
+		 *   with parts of the method name itself, not the arguments.
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/descriptor/phrases/SendPhraseDescriptor.kt:264]: Duplicate `@param` comment for `consumedTokens`.

File: avail/src/main/kotlin/avail/utility/Strings.kt [hunk 14]
```diff
@@ -373,4 +373,22 @@ object Strings
...
+
+	private val escapesForHTML = mapOf(
+		'&' to "&amp;",
+		'<' to "&lt;",
+		'>' to "&gt;",
+		'"' to "&quot;",
+		'\'' to "&#x27;",
+		'/' to "&#x2F;")
+
+	fun String.escapedForHTML() = buildString {
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/utility/Strings.kt:385]: Add doc comments.

File: avail/src/main/kotlin/avail/anvil/Styles.kt [hunk 15]
```diff
@@ -685,6 +688,56 @@ object StyleApplicator
...
+
+	/**
+	 * A [TokenStyle] contains information about which [PhraseNode] is
+	 * applicable for a span of source having this invisible style (under the
+	 * [PhraseNodeAttributeKey]), as well as which of the phrase's atom's
+	 * [MessageSplitter] parts occurred in this span of the source.
+	 */
+	data class TokenStyle(
+		val phraseNode: PhraseNode,
+		val tokenIndexInName: Int)
```
Comment (richATavail) [inline avail/src/main/kotlin/avail/anvil/Styles.kt:731]: Add `@property` comments for these.

**Review by richATavail — CHANGES_REQUESTED** [review]
I'm not familiar with how a lot of the work in this commit is put together, so I focused more on things I could identify as requiring attention in a more localized manner.

File: avail/src/main/kotlin/avail/anvil/Styles.kt [hunk 16]
```diff
@@ -653,26 +658,24 @@ object StyleApplicator
...
+				compositeStyles.computeIfAbsent(compositeStyleName) {
+					val style = addStyle(it, defaultStyle)
+					val styles = styleNames.mapNotNull(::getStyle)
+					if (styles.isNotEmpty())
+					{
+						val combined = styles.drop(1).fold(
+							StyleAspects(styles.first())
+						) { aspect, nextStyle ->
+							aspect + StyleAspects(nextStyle)
 						}
```
[the remark this answers is not in the document]
Comment (markATavail) [reply avail/src/main/kotlin/avail/anvil/Styles.kt:670]: You'll like this.  I realized this was a reduce pattern, so I was able to shrink it to:
```
						val combined = styles
							.map(::StyleAspects)
							.reduce(StyleAspects::plus)
```

File: avail/src/main/kotlin/avail/compiler/CompilationContext.kt [hunk 17]
```diff
@@ -1319,6 +1334,133 @@ class CompilationContext constructor(
...
+						token.start())
+					val pastEnd = surrogateIndexConverter.availIndexToJavaIndex(
+						token.pastEnd())
+					PhraseNodeToken(start, pastEnd, indexInName.extractInt)
+				}
+			if (atomName === null && phrase.phraseKind == SEND_PHRASE)
+			{
+				println("HERE")
+			}
+			//atomName = atomName ?: stringFrom("(${phrase.phraseKind.name})")
```
[the remark this answers is not in the document]
Comment (markATavail) [reply avail/src/main/kotlin/avail/compiler/CompilationContext.kt:1439]: Removed it, and simplified some earlier logic that's no longer needed.

File: avail/src/main/kotlin/avail/anvil/AvailEditor.kt [hunk 18]
```diff
@@ -482,19 +693,37 @@ class AvailEditor constructor(
 		panel.border = EmptyBorder(10, 10, 10, 10)
 		background = panel.background
 		val sourcePaneScroll = sourcePane.scrollTextWithLineNumbers()
+		val parseStructureListScroll = parseStructureList.scroll()
```
[the remark this answers is not in the document]
Comment (markATavail) [reply avail/src/main/kotlin/avail/anvil/AvailEditor.kt]: Moved it to its own window.

File: avail/src/main/kotlin/avail/utility/CollectionExtensions.kt [hunk 19]
```diff
@@ -424,3 +424,28 @@ fun <T1, T2, T3, T4, T5, T6, T7, T8, T9, T10> t(
...
+ */
+fun <X> X?.iterableWith(advance: (X)->X?): Iterable<X> = object : Iterable<X>
+{
+	override fun iterator() = object : Iterator<X>
+	{
+		var current = this@iterableWith
+		override fun hasNext() = current !== null
+		override fun next() = current!!.also { current = advance(it) }
+	}
+}
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/utility/CollectionExtensions.kt:451]: Neat utility function. This is the one that you mentioned to me earlier, I think.

File: avail/src/main/kotlin/avail/utility/ObjectTracer.kt [hunk 20]
```diff
@@ -161,9 +161,7 @@ constructor (
 	private fun gettersForClass(cls: Class<*>): List<(Any)->Any?> =
 		cacheByClass.computeIfAbsent(cls) { theClass ->
 			val getters = mutableListOf<(Any)->Any?>()
-			var eachCls: Class<*>? = theClass
-			while (eachCls !== null)
-			{
+			theClass.iterableWith { it.superclass }.forEach { eachCls ->
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/utility/ObjectTracer.kt:164]: Nice usage of the new utility also.

**Review by toddATavail — APPROVED** [review]
This looks incredibly good overall, and I'm going to merge it. But there are some clear bugs in `MessageSplitter` related to misuse of `apply`. @richATavail found one, and I found a second one. I might actually patch these up myself when I merge your branch into mine, but you should probably independently fix it anyway, in case I don't remember.

**Pull request merged**